### PR TITLE
[448] Build docker image

### PR DESCRIPTION
**Pull request by saliceti**
## What
Build the docker image, with or without cache and push to the Github container registry.

## How to review
Use as described in the README
Successful run: https://github.com/DFE-Digital/teacher-pay-calculator/actions/runs/5415001504/jobs/9843025806

### CODE REVIEW

File: build-docker-image/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+- `docker_repository`: Repository name in the container registry. e.g. ghcr.io/dfe-digital/register-trainee-teachers (Required)
+- `github_token`: Default Github token retrieved via secrets.github_token (Required)
+- `snyk_token`: Snyk token for image vulnerability scanning (Default: false)
+- `reuse_cache`: Use previously built docker image layers to improve build time. Set to false to refresh image (Default: true)
+
+## Example
+
+```yaml
+- name: Build and push docker image
+  uses: DFE-Digital/github-actions/build-docker-image@main
```
Comment (thomasleese) [inline build-docker-image/README.md:19]:
```suggestion
  uses: DFE-Digital/github-actions/build-docker-image@master
```

This repo hasn't been renamed yet - although I'd be in favour of doing that.

File: build-docker-image/action.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    required: true
+  github_token:
+    description: Default Github token retrieved via secrets.github_token
+    required: true
+  snyk_token:
+    description: Snyk token for image vulnerability scanning
+  reuse_cache:
+    description: Use previously built docker image layers to improve build time. Set to false to refresh image.
+    default: true
+    type: boolean
```
Comment (thomasleese) [inline build-docker-image/action.yml:16]: The convention with GitHub Actions (although not written down as far as I can tell, but [is shown in their examples](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs)) seems to be to use dashes in place of underscores with inputs and outputs. Not a blocker, but just thought it was worth sticking to one.

File: build-docker-image/action.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        build-args: COMMIT_SHA=${{ env.IMAGE_TAG }}
+
+    - name: Run Snyk to check Docker image for vulnerabilities
+      uses: snyk/actions/docker@master
+      if: inputs.snyk_token != ''
+      env:
+        snyk_token: ${{ inputs.snyk_token }}
+      with:
+        image: ${{ inputs.docker_repository }}:${{ env.IMAGE_TAG }}
+        args: --file=Dockerfile --severity-threshold=high --exclude-app-vulns
```
Comment (gunndabad) [inline build-docker-image/action.yml:79]: Worth having an input to override the working directory?

**Review by thomasleese — COMMENTED** [review]
Looks good, got two minor comments.

File: build-docker-image/action.yml (hunk 3 above)
Comment (saliceti) [reply build-docker-image/action.yml:79]: Do you have a use case for it or could we add it when necessary?

**Review by gunndabad — APPROVED** [review]

**Review by thomasleese — APPROVED** [review]

File: build-docker-image/action.yml (hunk 3 above)
Comment (saliceti) [reply build-docker-image/action.yml]: added input to override default

**Pull request closed without merging**